### PR TITLE
ST FSDEV EP0 left in stall

### PR DESCRIPTION
**Pull request by pigrew**
The ST FSDEV driver has been sometimes not completing USB enumeration (perhaps 10% of the time). I have not looked at it with a USB analyzer, but my feeling is that it is leaving the EP in a still condition after the host requests the device qualifier (which does not exist), and it's never leaving stall.

This patch removes some hardware validation of control status packets being 0-length, and also resets EP0 (both directions) to NAK upon receiving a SETUP packet.

This patch does not seem correct to me, but seems like it's working better than the original code. I'm not sure where the proper place to reset EP0 from stall is, or maybe this is the correct patch for that. I'll have to study the USB spec more, and if the hardware will automatically NAK control in/out after receiving a SETUP (even though the EP was set to stall).

This patch also adds a check to only pass setup packets to the stack if they are exactly 8 bytes long.

The hardware manual suggests setting the opposite direction of the transfer to stall during all but the last packet. It's not immediately clear what to do to the code, so I'm submitting this as an interim patch to improve the driver while thinking about the correct fix (perhaps in days, perhaps in weeks).

If desired, I can create an issue in the tracker to remind us to improve the driver to use EP_KIND and better stall/unstall.

### CODE REVIEW

**Comment by hathach** [comment]
@pigrew could you tell me steps to reproduce the issue, I will try to capture the usb to see what is the exact cause. Sometime we make changes and the issue goes away, but we may only fix 90% of the cause, and it may still re-occur in the future.

**Comment by pigrew** [comment]
It's hard to reproduce. It has been happening about 10% of the time while plugging in the USBTMC demo. WireShark shows that the device is stalling descriptor requests, so I'm fairly sure that it's due to a stall condition and/or the EP_KIND flag not being reset on EP0, after an error. We may be able to introduce the bug by sending an illegal control transaction (bad request number, etc), and then trying to make a proper request, to see if the stall goes away.

I won't be able to do that tonight, but maybe tomorrow I can get you a procedure to reproduce it.

Come to think of it, this would be a good test for other device drivers too.

**Comment by hathach** [comment]
> It's hard to reproduce. It has been happening about 10% of the time while plugging in the USBTMC demo. WireShark shows that the device is stalling descriptor requests, so I'm fairly sure that it's due to a stall condition and/or the EP_KIND flag not being reset on EP0, after an error. We may be able to introduce the bug by sending an illegal control transaction (bad request number, etc), and then trying to make a proper request, to see if the stall goes away.
> 
> I won't be able to do that tonight, but maybe tomorrow I can get you a procedure to reproduce it.
> 
> Come to think of it, this would be a good test for other device drivers too.

Anytime you feel convenient, though I think we don't need to generate an protocol error to test with the stack. USB stack is very established, all the hosts should follows that. It is a bit over engineered and may force us to write never-run code :) . I am happy with trouble-shooting when the bug arises :)

**Comment by pigrew** [comment]
Sorry, I've been consumed with babysitting antenna simulations, so have not had a chance to reply here.

I'm attaching a WireShark log, showing that the device is stalling every transaction. I don't have a way to consistently tigger the bug, just plug it in and unplug it a bunch of times with the USBTMC demo. I don't know if it also is triggered by other classes... have not tried them as much.
[stall_log.pcapng.gz](https://github.com/hathach/tinyusb/files/3663227/stall_log.pcapng.gz)

File: src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c [hunk 1]
```diff
@@ -419,10 +418,15 @@ static uint16_t dcd_ep_ctr_handler(void)
           uint8_t userMemBuf[8];
           /* Get SETUP Packet*/
           count = pcd_get_ep_rx_cnt(USB, EPindex);
-          //TU_ASSERT_ERR(count == 8);
-          dcd_read_packet_memory(userMemBuf, *pcd_ep_rx_address_ptr(USB,EPindex), 8);
+          if(count == 8) // Setup packet should always be 8 bits. If not, ignore it, and try again.
```
Comment (hathach) [inline src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c]: I guess you mean 8 bytes not bits. Le'ts me know when you think PR is ready for merging.

**Comment by pigrew** [comment]
@hathach , I'm happy with the patch now. I also added a testcase to the USBTMC test script in order to verify that EP0 can unstall.

Before the patch, the test fails. After the patch, it passes.

File: src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c (hunk 1 above)
Comment (pigrew) [reply src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c]: Yep, 64 bits. Fixed.

**Pull request merged**